### PR TITLE
tc qdisc: netem add support to duplicate packets.

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

**Comment by mrc0mmand** [comment]
```
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]: ==38784==ERROR: LeakSanitizer: detected memory leaks
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]: Direct leak of 7 byte(s) in 1 object(s) allocated from:
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #0 0x7fc2a73cf609 in strdup (/usr/lib/clang/9.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x69609)
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #1 0x5631a2434c38 in qdisc_configure /build/build/../src/network/tc/qdisc.c:136:28
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #2 0x5631a25116f2 in link_configure_qdiscs /build/build/../src/network/networkd-link.c:2596:21
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #3 0x5631a24ea292 in link_configure /build/build/../src/network/networkd-link.c:2616:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #4 0x5631a24f5532 in link_initialized_and_synced /build/build/../src/network/networkd-link.c:3028:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #5 0x5631a24efd10 in link_initialized_handler /build/build/../src/network/networkd-link.c:3038:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #6 0x7fc2a6e8fb38 in process_reply /build/build/../src/libsystemd/sd-netlink/sd-netlink.c:350:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #7 0x7fc2a6e8a8b5 in process_running /build/build/../src/libsystemd/sd-netlink/sd-netlink.c:417:21
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #8 0x7fc2a6e8a50c in sd_netlink_process /build/build/../src/libsystemd/sd-netlink/sd-netlink.c:446:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #9 0x7fc2a6e8d762 in time_callback /build/build/../src/libsystemd/sd-netlink/sd-netlink.c:728:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #10 0x7fc2a6ed1480 in source_dispatch /build/build/../src/libsystemd/sd-event/sd-event.c:2836:21
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #11 0x7fc2a6ed06f0 in sd_event_dispatch /build/build/../src/libsystemd/sd-event/sd-event.c:3241:21
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #12 0x7fc2a6ed2fcc in sd_event_run /build/build/../src/libsystemd/sd-event/sd-event.c:3299:21
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #13 0x7fc2a6ed38bd in sd_event_loop /build/build/../src/libsystemd/sd-event/sd-event.c:3321:21
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #14 0x5631a23d46d5 in run /build/build/../src/network/networkd.c:122:13
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #15 0x5631a23d3fe3 in main /build/build/../src/network/networkd.c:129:1
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]:     #16 0x7fc2a6516152 in __libc_start_main (/lib64/libc.so.6+0x27152)
Oct 30 20:22:04 arch.localdomain systemd-networkd[38784]: SUMMARY: AddressSanitizer: 7 byte(s) leaked in 1 allocation(s).


 ____________________________________________
/ Found   1 sanitizer errors (  1 ASan,   0  \
| UBSan,   0 MSan). Looks like you need to   |
\ look at the log                            /
 --------------------------------------------
 \
  \
     __
    /  \
    |  |
    @  @
    |  |
    || |/
    || ||
    |\_/|
    \___/
```

File: src/network/tc/qdisc.c [hunk 1]
```diff
@@ -132,25 +133,29 @@ int qdisc_configure(Link *link, QDiscs *qdisc) {
...
+                tca_kind = strdup("netem");
+                if (!tca_kind)
+                        return log_oom();
 
                 r = network_emulator_fill_message(link, qdisc, req);
                 if (r < 0)
                         return r;
         }
 
+        r = sd_netlink_message_append_string(req, TCA_KIND, tca_kind);
```
Comment (yuwata) [inline src/network/tc/qdisc.c]: `if (tca_kind) {`

File: src/network/tc/qdisc.c [hunk 2]
```diff
@@ -132,25 +133,29 @@ int qdisc_configure(Link *link, QDiscs *qdisc) {
...
-                r = sd_netlink_message_append_string(req, TCA_KIND, "clsact");
-                if (r < 0)
-                        return log_link_error_errno(link, r, "Could not append TCA_KIND attribute: %m");
         }
 
         if (qdisc->has_network_emulator) {
-                r = sd_netlink_message_append_string(req, TCA_KIND, "netem");
-                if (r < 0)
-                        return log_link_error_errno(link, r, "Could not append TCA_KIND attribute: %m");
+                tca_kind = strdup("netem");
```
Comment (yuwata) [inline src/network/tc/qdisc.c]: So, if `NetworkEmulatorLossRate=` is set, then `Parent=clsact` is partially ignored?

At least, this needs to be freed before `strdup()`, so, please use `free_and_strdup()`.

Comment (ssahani) [reply src/network/tc/qdisc.c]: We can add `clsact` like `sudo tc qdisc add dev veth99 clsact`
here no Netem conf is there but still then we able to create . By doing this we can replicate the same.  If Netem conf is available then we have to take that in consideration and apply Netem conf with clsact.

```
[TrafficControlQueueingDiscipline]
Parent=clsact
```

**Comment by ssahani** [comment]
updated thanks for review @yuwata .

**Comment by yuwata** [comment]
```
======================================================================
FAIL: test_bind_carrier (__main__.NetworkdNetworkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test-network/systemd-networkd-tests.py", line 1973, in test_bind_carrier
    self.assertRegex(output, 'UP,LOWER_UP')
AssertionError: Regex didn't match: 'UP,LOWER_UP' not found in '220: test1: <BROADCAST,NOARP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000\n    link/ether 2e:5c:e2:ab:4c:63 brd ff:ff:ff:ff:ff:ff'

----------------------------------------------------------------------
```
The test is flaky recently, not related.

**Pull request merged**